File: leetcode/p0127-word-ladder.py
#!/usr/bin/env python3
from time import time
from time import time_ns

class Solution(object):
    from typing import List
    #
    # TLE
    #
    def ladderLength1(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        if endWord not in wordList:
            return 0
        for word in wordList:
            if self.connected(beginWord, word):
                break
        queue = [beginWord]
        visited = set()
        res = 0
        while len(queue) > 0:
            res += 1
            tmp_queue = []
            for word in queue:
                if word == endWord:
                    return res
                visited.add(word)
                for next_word in wordList:  # slow
                    if next_word not in visited and self.connected(word, next_word):
                        tmp_queue.append(next_word)
            queue = tmp_queue
        return 0

    #
    # TLE
    #
    def ladderLength2(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        if endWord not in wordList:
            return 0
        for word in wordList:
            if self.connected(beginWord, word):
                break
        queue = [beginWord]
        nonvisited = set(wordList)
        res = 0
        while len(queue) > 0:
            res += 1
            tmp_queue = []
            for word in queue:
                if word == endWord:
                    return res
                nonvisited.discard(word)
                for next_word in nonvisited:
                    if self.connected(word, next_word):
                        tmp_queue.append(next_word)
            queue = tmp_queue
        return 0

    #
    # TLE
    #
    def ladderLength3(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        from collections import defaultdict
        if endWord not in wordList:
            return 0
        for word in wordList:
            if self.connected(beginWord, word):
                break
        queue = [beginWord]
        map = defaultdict(list)
        visited = set([beginWord])
        res = 0

        if beginWord not in wordList:
            wordList.append(beginWord)
        for i in range(len(wordList) - 1):
            for j in range(i + 1, len(wordList)):
                if self.connected(wordList[i], wordList[j]):
                    map[wordList[i]].append(wordList[j])
                    map[wordList[j]].append(wordList[i])
        while len(queue) > 0:
            print(queue)
            res += 1
            tmp_queue = []
            for word in queue:
                if word == endWord:
                    return res
                visited.add(word)
            for word in queue:
                for next_word in map[word]:
                    if next_word not in visited:
                        visited.add(next_word)
                        tmp_queue.append(next_word)
            queue = tmp_queue
        return 0

    #
    # TLE
    # but Java version can pass
    #
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        if endWord not in wordList:
            return 0
        for word in wordList:
            if self.connected(beginWord, word):
                break
        wordList.append(beginWord)

        cnt = 0
        graph = []
        for i in range(len(wordList)):
            graph.append([])

        for i in range(len(wordList)-1):
            s = time()
            for j in range(i+1, len(wordList)):
                if self.connected(wordList[i], wordList[j]):
                    cnt += 1
                    graph[i].append(j)
                    graph[j].append(i)

        res = 0
        target = wordList.index(endWord)
        queue = [len(wordList) - 1]
        visited = [False] * len(wordList)
        while len(queue) > 0:
            res += 1
            tmp_queue = []
            for index in queue:
                if index == target:
                    return res
                visited[index] = True
            for index in queue:
                for next_index in graph[index]:
                    if visited[next_index] is False:
                        visited[next_index] = True
                        tmp_queue.append(next_index)
            queue = tmp_queue
        return 0

    def connected(self, word1, word2) -> bool:
        cnt = 0
        for i in range(len(word1)):
            if cnt > 1:
                return False
            if word1[i] != word2[i]:
                cnt += 1
        return cnt == 1

if __name__ == '__main__':
    assert Solution().ladderLength("hit", "cog", ["hot","dot","dog","lot","log","cog"]) == 5
    assert Solution().ladderLength("hit", "cog", ["hot","dot","dog","lot","log"]) == 0
    assert Solution().ladderLength("leet", "code", ["lest","leet","lose","code","lode","robe","lost"]) == 6

File: leetcode/rust_leetcode/src/bin/p0127_word_ladder.rs

fn main() {}

struct Solution;

impl Solution {
    pub fn ladder_length(begin_word: String, end_word: String, word_list: Vec<String>) -> i32 {
        use std::collections::{HashMap, HashSet};
        use std::time::SystemTime;

        if !word_list.contains(&end_word) {
            return 0;
        }
        for word in word_list.iter() {
            if Solution::connected(begin_word.as_str(), word.as_str()) {
                break
            }
        }

        let mut word_list = word_list;
        word_list.push(begin_word.clone());

        let x = SystemTime::now();
        let mut graph = vec![vec![];word_list.len()];
        for i in 0..word_list.len() - 1{
            for j in i + 1..word_list.len() {
                if Solution::connected(word_list[i].as_str(), word_list[j].as_str()) {
                    graph[i].push(j);
                    graph[j].push(i);
                }
            }
        }

        let mut res = 0;
        let target = word_list.iter().position(|x| x == &end_word).unwrap();
        let mut visited = vec![false; word_list.len()];
        let mut queue = vec![word_list.len() - 1];
        while !queue.is_empty() {
            res += 1;
            for &index in queue.iter() {
                if index == target {
                    return res;
                }
                visited[index] = true;
            }
            let indices: Vec<_> = queue.drain(..).collect();
            for index in indices {
                for &next_index in graph[index].iter() {
                    if !visited[next_index] {
                        visited[next_index] = true;
                        queue.push(next_index);
                    }
                }
            }
        }
        return 0;
    }

    fn connected(word1: &str, word2: &str) -> bool {
        let mut cnt = 0;
//        let chars1: Vec<_> = word1.chars().collect(); // very slow
        let bytes1 = word1.as_bytes();
        let bytes2 = word2.as_bytes();
        for i in 0..bytes1.len() {
            if cnt > 1 {
                return false
            }
            if bytes1[i] != bytes2[i] {
                cnt += 1;
            }
        }
        cnt == 1
    }
}

#[cfg(test)]
mod tests {

    use super::*;

    #[test]
    fn test1() {
        let v1 = vec!["hot","dot","dog","lot","log","cog"].iter().map(|&w| w.to_string()).collect();
        assert_eq!(Solution::ladder_length("hit".to_string(), "cog".to_string(), v1), 5);

        let v2 = vec!["hot","dot","dog","lot","log"].iter().map(|&w| w.to_string()).collect();
        assert_eq!(Solution::ladder_length("hit".to_string(), "cog".to_string(), v2), 0);

        let v3 = vec!["lest","leet","lose","code","lode","robe","lost"].iter().map(|&w| w.to_string()).collect();
        assert_eq!(Solution::ladder_length("leet".to_string(), "code".to_string(), v3), 6);
    }
}